modules/javafx.controls/src/main/java/javafx/scene/control/skin/MenuBarSkin.java

```diff
@@ -389,62 +276,141 @@ public MenuBarSkin(final MenuBar control) {
            acceleratorKeyCombo = KeyCombination.keyCombination("F10");
         }
 
-        altKeyEventHandler = e -> {
-            if (e.getEventType() == KeyEvent.KEY_PRESSED) {
-                // Clear menu selection when ALT is pressed by itself
-                altKeyPressed = false;
-                if (e.getCode() == ALT && !e.isConsumed()) {
-                    if (focusedMenuIndex == -1) {
-                        altKeyPressed = true;
-                    }
-                    unSelectMenus();
-                }
-            } else if (e.getEventType() == KeyEvent.KEY_RELEASED) {
-                // Put focus on the first menu when ALT is released
-                // directly after being pressed by itself
-                if (altKeyPressed && e.getCode() == ALT && !e.isConsumed()) {
-                    firstMenuRunnable.run();
-                }
-                altKeyPressed = false;
-            }
-        };
-        weakSceneAltKeyEventHandler = new WeakEventHandler<>(altKeyEventHandler);
-
-        Utils.executeOnceWhenPropertyIsNonNull(control.sceneProperty(), (Scene scene) -> {
-            scene.getAccelerators().put(acceleratorKeyCombo, firstMenuRunnable);
-            scene.addEventHandler(KeyEvent.ANY, weakSceneAltKeyEventHandler);
-        });
-
         ParentTraversalEngine engine = new ParentTraversalEngine(getSkinnable());
         engine.addTraverseListener((node, bounds) -> {
             if (openMenu != null) openMenu.hide();
             setFocusedMenuIndex(0);
         });
         ParentHelper.setTraversalEngine(getSkinnable(), engine);
 
-        control.sceneProperty().addListener((ov, t, t1) -> {
-            // remove event handlers / filters from the old scene (t)
-            if (t != null) {
-                if (weakSceneKeyEventHandler != null) {
-                    t.removeEventFilter(KeyEvent.KEY_PRESSED, weakSceneKeyEventHandler);
-                }
-                if (weakSceneMouseEventHandler != null) {
-                    t.removeEventFilter(MouseEvent.MOUSE_CLICKED, weakSceneMouseEventHandler);
-                }
-                if (weakSceneAltKeyEventHandler != null) {
-                    t.removeEventHandler(KeyEvent.ANY, weakSceneAltKeyEventHandler);
-                }
+        lh.addChangeListener(control.sceneProperty(), true, (scene) -> {
+            if (sceneListenerHelper != null) {
+                sceneListenerHelper.disconnect();
+                sceneListenerHelper = null;
             }
 
-            /**
-             * remove the f10 accelerator from the old scene
-             * add it to the new scene
-             */
-            if (t != null) {
-                t.getAccelerators().remove(acceleratorKeyCombo);
-            }
-            if (t1 != null ) {
-                t1.getAccelerators().put(acceleratorKeyCombo, firstMenuRunnable);
+            if (scene != null ) {
+                sceneListenerHelper = new ListenerHelper(MenuBarSkin.this);
```

Comment:
Why does this (still) need to rely on a weak reference?  Skins have a well specified life cycle do they not?

Reply:
You are right: some weak listeners remain, I did not want to re-write the whole thing for a fear of introducing regression and to keep the changes to a minimum.

The second `ListenerHelper` (line 293) gets disconnected in dispose(), or when the skin is collected (since the skin may not be explicitly uninstalled, but instead the whole component or `Pane` might be removed from the scene and discarded.

Reply:
> but instead the whole component or Pane might be removed from the scene and discarded.

If that happens, weak reference won't make a difference there either. Removing an entire branch from the Scene and not referring to it anymore will do correct clean-up even without a call to dispose and without the use of weak references.

The only use case I see is that we still don't trust the Skin lifecycle to be adhered to and "just in case" are using weak references in case somehow dispose is not called. I would much prefer to see an actual reason to use it, and then adding a comment as to why this weak reference is needed so that in 6 months time we're not scratching our heads as to why a weak reference is needed here.

Reply:
just to be sure, which weak listener are you referring to?

Reply:
I didn't mean a weak listener, but a weak reference.  This line creates one:

     sceneListenerHelper = new ListenerHelper(MenuBarSkin.this);

I think a plain `new ListenerHelper()` would work correct here.  If you don't think so, I think it may be good to add a comment why the weak reference is needed to make the clean up work correctly.

Reply:
I have to agree with you - it is not needed in this case.  The reason I put it there was "what if Stage.hide() gets called and the whole thing disappears, perhaps we want to make sure that all listeners that were registered via `sceneListenerHelper` get disconnected.  The thing is, all these listeners are registered against the scene belonging to that stage, so we are ok.